metricbeat/module/vsphere/cluster/cluster.go

```diff
@@ -0,0 +1,177 @@
+func init() {
+	mb.Registry.MustAddMetricSet("vsphere", "cluster", New,
+		mb.WithHostParser(vsphere.HostParser),
+		mb.DefaultMetricSet(),
+	)
+}
+
+// MetricSet type defines all fields of the MetricSet.
+type ClusterMetricSet struct {
+	*vsphere.MetricSet
+}
+
+type assetNames struct {
+	outputNtNames []string
+	outputDsNames []string
+	outputHsNames []string
+}
+
+// New creates a new instance of the MetricSet.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	ms, err := vsphere.NewMetricSet(base)
+	if err != nil {
+		return nil, err
+	}
+	return &ClusterMetricSet{ms}, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *ClusterMetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	client, err := govmomi.NewClient(ctx, m.HostURL, m.Insecure)
+	if err != nil {
+		return fmt.Errorf("error in NewClient: %w", err)
+	}
+	defer func() {
+		if err := client.Logout(ctx); err != nil {
+			m.Logger().Debug(fmt.Errorf("error trying to logout from vshphere: %w", err))
+		}
+	}()
+
+	c := client.Client
+
+	// Create a view of Cluster objects
+	mgr := view.NewManager(c)
+
+	v, err := mgr.CreateContainerView(ctx, c.ServiceContent.RootFolder, []string{"ClusterComputeResource"}, true)
+	if err != nil {
+		return fmt.Errorf("error in CreateContainerView: %w", err)
+	}
+
+	defer func() {
+		if err := v.Destroy(ctx); err != nil {
+			m.Logger().Errorf("error trying to destroy view from vshphere: %w", err)
+		}
+	}()
+
+	// Retrieve summary property for all Clusters
+	var clt []mo.ClusterComputeResource
+	err = v.Retrieve(ctx, []string{"ClusterComputeResource"}, []string{}, &clt)
+	if err != nil {
+		return fmt.Errorf("error in Retrieve: %w", err)
+	}
+
+	pc := property.DefaultCollector(c)
+	for i := range clt {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			assetNames, err := getAssetNames(ctx, pc, &clt[i])
+			if err != nil {
+				m.Logger().Errorf("Failed to retrieve object from cluster %s: %v", clt[i].Name, err)
+			}
+
+			if clt[i].Configuration.DasConfig.AdmissionControlEnabled == nil {
+				m.Logger().Warnf("Metric das_config.admission.control.enabled not found")
+			}
+			if clt[i].Configuration.DasConfig.Enabled == nil {
+				m.Logger().Warnf("Metric das_config.enabled not found")
+			}
+
+			reporter.Event(mb.Event{
+				MetricSetFields: m.mapEvent(clt[i], assetNames),
+			})
+		}
+	}
+	return nil
+
+}
+
+func getAssetNames(ctx context.Context, pc *property.Collector, cl *mo.ClusterComputeResource) (*assetNames, error) {
+	referenceList := make([]types.ManagedObjectReference, 0, len(cl.Datastore)+len(cl.Host))
+	referenceList = append(referenceList, cl.Datastore...)
+	referenceList = append(referenceList, cl.Host...)
```

Comment:
```suggestion
	referenceList := append(cl.Datastore..., cl.Host..)
```

Reply:
@kush-elastic two appends are required. Both are variadic (...) but only the last argument can be this.

We can although use of append inside of an append but the current implementation is clear. Here, we are appending 2 slices to another slice.

Reply:
Oh, right. but this will do the trick
```suggestion
	referenceList := append(cl.Datastore, cl.Host...)
```

Reply:
Yes!